### PR TITLE
Improve NumericInput and AbstractButton internals

### LINKED ISSUE #4200: NumericInput events run incorrectly when using min or max props
#### Environment

- __Package version(s)__: 3.28 / 3.29
- __Operating System__: Linux
- __Browser name and version__:  Firefox 77 / Chromium 81

#### Code Sandbox

Link to a minimal repro (fork [this code sandbox](https://codesandbox.io/s/blueprint-sandbox-et9xy)): https://codesandbox.io/s/blueprint-numericinput-vxkuu

#### Steps to reproduce

1. Hold arrow button until limit is reached, then do the same with opposite button, repeating a couple of times will
   make all events to not work properly (button clicks set more than 1 step at once, and keyboard up and down arrow don't stop until limit is reached, even if tapped once)

#### Actual behavior

Button & keyboard events don't work properly

#### Expected behavior

Behaviour shouldn't not change when usin min/max

Edit: Seems I found the cause of the issue, this bug doesn't happen if you release the mouse before reaching the limit, or upon reaching the limit, outside of the disabled button. I've checked the source and it seems stopContinuousChange is not firing when releasing the mouse on the disabled button

### PR DESCRIPTION
**Pull request by ejose19**
#### Checklist

- [X] Includes tests
- [X] Update documentation (No changes to docs)

#### Changes proposed in this pull request:

- ~~Add a check to each iteration of handleContinuousChange, since mouseup event won't fire if the mouse is released on the disabled element (after it reached min/max limit) current behaviour keeps running it indefinitely when using min/max props and releasing the mouseup after described event.~~ Fixed using AnchorButton

- ~~Allow accessing class instance during onValueChange callback, this is necessary as if doing any conditional logic to allow a value, one needs to update the internal value to keep it in sync with given value prop (see [my comment on this issue for more details](https://github.com/palantir/blueprint/pull/4131#issuecomment-651435032))~~ Removed, another PR improves controlled mode

- Add ~~componentDidUpdate~~ getDerivedStateFromProps to AbstractButton, related to the 1st issue, its handleKeyUp is never fired if the button is disabled, so internal state is still "active" while the button is disabled (which is not correct) displaying inconsistent UI to the user when using along other components (For example NumericInput)

#### Reviewers should focus on:

Mentioned changes

#### Screenshot

N/A

### CODE REVIEW

**Comment by palantirtech** [comment]
Thanks for your interest in [palantir/blueprint](https://github.com/palantir/blueprint), @ejose19! Before we can accept your pull request, you need to sign our contributor license agreement - just visit https://cla.palantir.com/ and follow the instructions. Once you sign, I'll automatically update this pull request.

File: packages/core/src/components/button/abstractButton.tsx [hunk 1]
```diff
@@ -134,6 +134,16 @@ export abstract class AbstractButton<H extends React.HTMLAttributes<HTMLElement>
         };
     }
 
+    // A disabled element cannot be active, there're are cases where
```
Comment (adidahiya) [inline packages/core/src/components/button/abstractButton.tsx]:
```suggestion
    // A disabled element cannot be active, but there are cases where
```

File: packages/core/src/components/button/abstractButton.tsx [hunk 2]
```diff
@@ -134,6 +134,16 @@ export abstract class AbstractButton<H extends React.HTMLAttributes<HTMLElement>
...
+    // this situation can happen (NumericInput buttons), hence the need
+    // to check if disabled prop has been passed and change isActive
+    // accordingly
+    public componentDidUpdate() {
```
Comment (adidahiya) [inline packages/core/src/components/button/abstractButton.tsx]: I think this should be implemented inside `static getDerivedStateFromProps` instead:

```ts
    protected static getDerivedStateFromProps(props: IButtonProps, state: IButtonState) {
        if (state.isActive && props.disabled) {
            return {
                isActive: false,
            };
        }

        return null;
    }
```

**Review by adidahiya — CHANGES_REQUESTED** [review]
> Allow accessing class instance during onValueChange callback, this is necessary as if doing any conditional logic to allow a value, one needs to update the internal value to keep it in sync with given value prop

I'm not yet convinced that this is necessary. What kind of conditional logic are you doing? Can we support that use case some other way instead of providing access to the component's internals? I don't want to increase the size of the public API this way.

---

Also, I think this whole issue might be easier solved by simply replacing the `Button` components rendered by `NumericInput` with `AnchorButton` components... they are able to receive events when `disabled=true`

**Comment by ejose19** [comment]
@adidahiya Regarding the instance access, if there's any other way then great, but I would like to know how you would do that. Here's the issue: Controlled mode, my value is set to 1, and NumericInput sets its internal state to '1', now I only want integers in a range, so I add a conditional logic to onValueChange callback to apply the new value to my state,  this theorically works good if the user is either using the buttons or tapping only numbers, but there are 2 cases: (pressing the letter 'e' is allowed, and pasting content into the field) first action is allowed and internal value is stored as '1e' but my state rejected it and it's 1, now they're not in sync and buttons won't display limits properly. On paste if text pasted is invalid (letters only), then internal state is VALUE_EMPTY and my state is still 1. So either we need to add a callback function like ("isValid") to determine if internal state should change, or as I'm  doing in this patch, let me change the internal state manually after I do my checks.

File: packages/core/src/components/button/abstractButton.tsx (hunk 2 above)
Comment (ejose19) [reply packages/core/src/components/button/abstractButton.tsx]: If this is better than componentDidUpdate then no issues, I will change it.

**Review by ejose19 — COMMENTED** [review]
Applied both recommendations and I confirm they work (Using AnchorButton on numericInput and getDerivedStateFromProps on abstractButton)

**Comment by adidahiya** [comment]
@ejose19 regarding controlled mode... check out the PR I just opened #4211... that should get you sorted. For this PR please just stick to fixing #4200. I'll merge my PR today and then you can test it.

File: packages/core/src/components/forms/numericInput.tsx [hunk 3]
```diff
@@ -151,7 +151,12 @@ export interface INumericInputProps extends IIntentProps, IProps {
     onButtonClick?(valueAsNumber: number, valueAsString: string): void;
 
     /** The callback invoked when the value changes due to typing, arrow keys, or button clicks. */
-    onValueChange?(valueAsNumber: number, valueAsString: string, inputElement: HTMLInputElement | null): void;
+    onValueChange?(
+        valueAsNumber: number,
+        valueAsString: string,
+        inputElement: HTMLInputElement | null,
+        instance: NumericInput,
```
Comment (adidahiya) [inline packages/core/src/components/forms/numericInput.tsx]: please revert this change unrelated to #4200

**Review by adidahiya — CHANGES_REQUESTED** [review]

File: packages/core/src/components/button/abstractButton.tsx (hunk 2 above)
Comment (adidahiya) [reply packages/core/src/components/button/abstractButton.tsx]: It's better because it's a more natural place to manage state updates, it gets called before `render`. Check out the React lifecycle documentation to see what kinds of behavior is recommended in the various methods

File: packages/core/src/components/forms/numericInput.tsx (hunk 3 above)
Comment (ejose19) [reply packages/core/src/components/forms/numericInput.tsx]: Tried your fork and apparently it works correctly, so I'm removing instance from this PR and will try again on official version once this and the other PR is merged

**Review by adidahiya — APPROVED** [review]

**Comment by adidahiya** [comment]
thanks for testing my branch. I'll merge yours first so you don't have to deal with conflicts

**Comment by ejose19** [comment]
@adidahiya Since there's a side effect caused by using AnchorButton, should we use the original idea instead?

> Add a check to each iteration of handleContinuousChange, since mouseup event won't fire if the mouse is released on the disabled element (after it reached min/max limit) current behaviour keeps running it indefinitely when using min/max props and releasing the mouseup after described event.

**Comment by adidahiya** [comment]
@ejose19 ok, yeah, I would be open to a change that switches to the original idea using Button instead
